### ISSUE TITLE
separate trainer, optimizer and loss

### ISSUE THREAD

**Issue by vzaguskin** [vzaguskin/machinelearne-rs#1]
see docs/adr/0001-separate-trainer-losses.md
## Context
The initial implementation has training loop as part of the model. We want generalization for loss functions, regularization methods and training strategies

## Decision
The proposal is to make separate structures:
    Loss function(MSE/MAE/BCE/ with or without regularization). Than will hardcode gradient formula for now.
    Optimizer - SGD to begin with, with learning rate making update step
    Trainer - with batch_size, training loop, stopping criterion
    Metrics
    Logging
